autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/service/AutofillProviderFillSuggestionActivity.kt

```diff
@@ -0,0 +1,200 @@
+            finish()
+            return
+        }
+        observeViewModel()
+    }
+
+    private fun observeViewModel() {
+        viewModel.commands()
+            .flowWithLifecycle(lifecycle, Lifecycle.State.STARTED)
+            .onEach { processCommand(it) }
+            .launchIn(lifecycleScope)
+    }
+
+    private fun processCommand(command: Command) {
+        when (command) {
+            is RequestAuthentication -> {
+                Timber.i("DDGAutofillService auth IS REQUIRED!")
+                deviceAuthenticator.authenticate(this) {
+                    when (it) {
+                        Success -> {
+                            viewModel.onUserAuthenticatedSuccessfully()
+                        }
+
+                        UserCancelled -> {
+                            finish()
+                        }
+
+                        is Error -> {
+                            finish()
+                        }
+                    }
+                }
+            }
+
+            ContinueWithoutAuthentication -> {
+                Timber.i("DDGAutofillService ContinueWithoutAuthentication credentialId: $credentialId")
+                credentialId?.let { nonNullId ->
+                    viewModel.continueAfterAuthentication(nonNullId)
+                } ?: run {
+                    finish()
+                }
+            }
+
+            is AutofillLogin -> {
+                autofillLogin(command.credentials)
+            }
+
+            ForceFinish -> finish()
+        }
+    }
+
+    fun autofillLogin(credential: LoginCredentials) {
+        val structure = assistStructure ?: return
+        val parsedNodes = autofillParser.parseStructure(structure)
+        val detectedNode: Pair<AutofillRootNode, ParsedAutofillField>? = parsedNodes.firstNotNullOfOrNull { node ->
+            val focusedDetectedField = node.parsedAutofillFields
+                .firstOrNull { field ->
+                    field.originalNode.isFocused && field.type != UNKNOWN
+                }
+            if (focusedDetectedField != null) {
+                return@firstNotNullOfOrNull Pair(node, focusedDetectedField)
+            }
+            val firstDetectedField = node.parsedAutofillFields.firstOrNull { field -> field.type != UNKNOWN }
+            if (firstDetectedField != null) {
+                return@firstNotNullOfOrNull Pair(node, firstDetectedField)
+            }
+            return@firstNotNullOfOrNull null
+        }
+
+        if (detectedNode == null) {
+            return
+        }
+
+        val fields = detectedNode.first.parsedAutofillFields.filter { it.type != UNKNOWN }
+
+        val dataset = buildDataset(fields, credential)
+
+        val resultIntent = Intent().apply {
+            putExtra(AutofillManager.EXTRA_AUTHENTICATION_RESULT, dataset)
+        }
+        this.setResult(Activity.RESULT_OK, resultIntent)
+        this.finish()
+    }
+
+    private fun buildDataset(
```

Comment:
Same non-blocking suggestion here:

- the whole of `buildDataset()` is duplicated across two places, non trivial and not easy to test 
- extract as a class ?